File: Programmers/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String, Integer> rankMap = new HashMap<>();

        for (int i = 0; i < players.length; i++) {
            rankMap.put(players[i], i);
        }

        for (String call : callings) {
            int i = rankMap.get(call);

            String frontPlayer = players[i - 1];
            players[i - 1] = call;
            players[i] = frontPlayer;

            rankMap.put(call, i - 1);
            rankMap.put(frontPlayer, i);
        }

        return players;
    }
}

File: Programmers/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    unordered_map<string, int> rank;
    
    for (int i = 0; i < (int)players.size(); i++) 
    {
        rank[players[i]] = i;
    }

    for (const string& name : callings) 
    {
        int curIdx = rank[name];
        int frontIdx = curIdx - 1;

        string frontPlayer = players[frontIdx];
        swap(players[curIdx], players[frontIdx]);

        rank[name] = frontIdx;
        rank[frontPlayer] = curIdx;
    }

    return players;
}